File: javascript/json-transform/src/functions/TransformerFunctionTypeOf.ts
import TransformerFunction from "./common/TransformerFunction";
import FunctionContext from "./common/FunctionContext";
import { isNullOrUndefined } from "../JsonHelpers";
import BigNumber from "bignumber.js";

const ACCEPTED_TYPES = new Set(["string", "boolean", "number"]);

class TransformerFunctionTypeOf extends TransformerFunction {
  constructor() {
    super({});
  }

  override async apply(context: FunctionContext): Promise<any> {
    const value = await context.getJsonElement(null);
    if (isNullOrUndefined(value)) {
      return "null";
    }
    const type = typeof value;
    if (ACCEPTED_TYPES.has(type)) {
      return type;
    }
    if (BigNumber.isBigNumber(value) || type === "bigint") {
      return "number";
    }
    if (Array.isArray(value)) {
      return "array";
    }
    return "object";
  }
}

export default TransformerFunctionTypeOf;

File: java/json-transform/src/main/java/ai/bizone/jsontransform/functions/TransformerFunctionTypeOf.java
package ai.bizone.jsontransform.functions;

import ai.bizone.jsontransform.functions.common.*;

import java.util.Map;

public class TransformerFunctionTypeOf extends TransformerFunction {

    public TransformerFunctionTypeOf() {
        super();
    }

    @Override
    public Object apply(FunctionContext context) {
        var adapter = context.getAdapter();
        var value = context.getJsonElement(null);
        if (value == null || adapter.isNull(value)) return "null";
        if (adapter.isJsonString(value)) return "string";
        if (adapter.isJsonNumber(value)) return "number";
        if (adapter.isJsonBoolean(value)) return "boolean";
        if (adapter.isJsonArray(value)) return "array";
        return "object";
    }
}
